Improved docstring and return type hints for `to_datetime`

smarie opened the pull request:
- [x] closes https://github.com/pandas-dev/pandas/issues/17973

Improved (still not ideal) fix for #42229 (datetime parsing behaviour in case of multiple time offsets due to daylight-savings in timezones) following https://github.com/pandas-dev/pandas/pull/42244 

See also https://github.com/pandas-dev/pandas/issues/42229#issuecomment-870009579

smarie committed:
Update datetimes.py

Following https://github.com/pandas-dev/pandas/pull/42244 , improved documentation about datetime parsing.

See also https://github.com/pandas-dev/pandas/issues/42229#issuecomment-870009579

jreback commented on pandas/core/tools/datetimes.py:
```diff
@@ -763,7 +763,7 @@ def to_datetime(
 
         - list-like:
-            - DatetimeIndex, if timezone naive or aware with the same timezone
+            - DatetimeIndex, if timezone naive or aware with constant time offset
```
how about leave the original & add this as well

smarie replied:
I am not sure to understand your comment @jreback , sorry about this. Could you be more precise please ? Do you mean a sentence such as 

```suggestion
            - DatetimeIndex, if timezone naive or timezone-aware with either
              constant time offset or constant timezone name.
```

?

mroeschke replied:
I get what you're trying to convey here (a timezone can have different UTC offsets), but I think the current language is fine.

I think a better contribution here would be to add a `Notes` section referencing pandas timezone handling so users can learn more about these subtleties. https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html#time-zone-handling

smarie replied:
Thanks for the suggestion @mroeschke ! Sorry for not having had the bandwidth to respond earlier.

The current docstring opposes two situations:

 - DatetimeIndex \[are returned\] if \[input datetimes are\] timezone naive or aware with the same timezone
 - Index of object dtype \[are returned\] if \[input datetimes are\] timezone aware with mixed time offsets

The first bullet is incorrect or at least very misleading, because it conflicts with the second bullet. Indeed datetimes that are "aware with the same timezone" **can** contain mixed time offsets. My proposal removes the conflict:

 - DatetimeIndex \[are returned\] if \[input datetimes are\] timezone naive or aware with **constant time offset**
 - Index of object dtype \[are returned\] if \[input datetimes are\] timezone aware with mixed time offsets

Does this make sense ?

Now concerning your suggestion for adding a "Notes" section, I think that it is great ! I'll include this link somewhere

smarie commented:
@jreback do you have any additional response to the above comment ? otherwise we can probably proceed with merging "as is"

jreback commented:
closing as stale, if you want to continue working, please ping.

Commit:
from code review: improved utc doc

Commit:
Merge branch 'master' of https://github.com/pandas-dev/pandas into patch-1

smarie commented:
@jreback thanks for the ping. I made changes in my branch but I do not see them here - I guess that you'll have to open the PR to see them ? I tried to update the documentation of the "utc" parameter as it is the main cause of all difficulties here.

smarie commented:
Thanks @mroeschke ! I'll push a few doc mods again and let you know.

Commit:
Improved overall readability by

 - adding an extended summary with details, instead of putting all details in the arguments and outputs description.
 - Fixing return type hints to add the type hints for datetime.datetime scalars or array-like.
 - adding an explicit "raises" section
 - adding illustrative examples

Commit:
minor improvement

smarie commented:
@jreback and @mroeschke  : all set.

Instead of trying to squeeze a complex explanation in the arguments and outputs definitions, I rather created a proper extended summary.

I also fixed the return type hints and added several examples. These examples show that the current behaviour is quite complex, but at least now users can see it :)

Commit:
Minor fix and improvement again

Commit:
Changed order of output description to match the global section doc

Commit:
Removed the "type: ignore" since the return type hints are now fixed

Commit:
Removed type hint-related mods (will move to a separate pr)

Commit:
Merge branch 'master' of https://github.com/pandas-dev/pandas into patch-1

smarie commented:
Thanks @mroeschke for the review ! I'll have a look when possible, probably next week.

Commit:
Removed backslash characters from doctests as per code review

Commit:
As per code review: replaced all "tz-" with "timezone-"

Commit:
Code review: capitalized if

mroeschke commented:
`to_datetime` has a lot of arguments and accepts a lot of array-like and scalar-like argument so naturally this function will be quite complex (as dates are quite complex) and ideally ideally all the interactions of inputs + arguments should be documented.

You are welcome to submit and issue per behavior you think needs changing, but note they may likely be API changes and may be difficult to deprecate + remove as other users may have been relying on this behavior for a while.

jreback reviewed (changes requested):
will look soon

jreback reviewed (changes requested):
can you organize the examples, there is a lot going on and its hard to tell what is what.

maybe put some comments to denote different sections of the examples.

you don't need to show every special case

Commit:
Compressed output description as per code review.

Commit:
Moved the general summary to a notes section

smarie committed:
Update pandas/core/tools/datetimes.py

smarie commented on pandas/core/tools/datetimes.py:
```diff
@@ -689,6 +689,47 @@ def to_datetime(
     """
     Convert argument to datetime.
 
+    This function converts a scalar, array-like, :class:`Series` or
```
Done

Commit:
As per code review: reduced the utc param description and added structure to the examples section.

smarie commented:
Hi @mroeschke  and @jreback , I took a few hours again to update this PR.

 - I reduced the comments about the utc parameter and replaced them with a link to the examples section
 - I made several sub-sections in the examples
 - I made the examples section look nicer
 - I tested my documentation with local build `python make.py --single pandas.to_datetime` : it looks quite good ! 

I might be over-optimistic here, but this could be actually matching your expectations :)

I let you have a look (please look at the locally built doc, it is far easier to follow) and tell me what you think !
Have a nice xmas break!

Commit:
Merge branch 'master' of https://github.com/pandas-dev/pandas into patch-1

 Conflicts:
	pandas/core/tools/datetimes.py

smarie commented on pandas/core/tools/datetimes.py:
```diff
@@ -723,13 +764,40 @@ def to_datetime(
     utc : bool, default None
-        Return UTC DatetimeIndex if True (converting any tz-aware
-        datetime.datetime objects as well).
+        Control timezone-related parsing, localization and conversion.
```
I moved the details to the examples section and added a link.

Commit:
Minor edits

smarie commented on pandas/core/tools/datetimes.py:
```diff
@@ -850,16 +927,78 @@ def to_datetime(
+    Special case: mixing timezone-aware string and datetime fails when
+    utc=False, even if they have the same time offset.
+
+    >>> from datetime import datetime, timezone, timedelta
```
I restructured the examples section, this should be now easier to follow for readers.

mroeschke commented on pandas/core/tools/datetimes.py:
```diff
@@ -726,13 +729,30 @@ def to_datetime(
+        - If ``True``, the function *always* returns a timezone-aware
+          UTC-localized Timestamp, Series or DatetimeIndex. To do this,
+          timezone-naive inputs are *localized* as UTC (e.g.
+          ``01-01-2020 01:00:00`` becomes ``01-01-2020 01:00:00Z``), while
```
I think the `e.g.` examples here should be includes in the `Examples` section to simplify the description here.

smarie replied:
ok

mroeschke commented on pandas/core/tools/datetimes.py:
```diff
@@ -726,13 +729,30 @@ def to_datetime(
+          timezone-aware inputs are *converted* to UTC (e.g.
+          ``01-01-2020 01:00:00+0100`` becomes ``01-01-2020 00:00:00Z``).
+
+        - If ``False`` (default), the result is a "best effort automation",
```
This description doesn't seem to describe much IMO.

How about `inputs will not be attempted to be explicitly coerced to UTC`

smarie replied:
Indeed this was a bit clunky. Leveraging your proposal, I wrote `inputs will not be coerced to UTC` but changed a bit the details.

mroeschke commented on pandas/core/tools/datetimes.py:
```diff
@@ -772,28 +792,81 @@ def to_datetime(
-        return will have datetime.datetime type (or corresponding
-        array/Series).
+        Return type depends on input (types in parenthesis correspond to
+        timezone or out-of-range timestamp handling issues):
```
`handling issue` -> `fallback handling`

smarie replied:
indeed the sentence was clunky. I feel that both aspects (timezone issues + out of range timestamps) need different names. What about :

    `types in parenthesis correspond to fallback in case of timezone issues or out-of-range timestamps`

mroeschke replied:
That's a better phrasing. Would replace `issues` with `unsuccessful timezone or out-of-range timestamp parsing`

smarie replied:
Nice. done.

mroeschke commented on pandas/core/tools/datetimes.py:
```diff
@@ -772,28 +792,81 @@ def to_datetime(
     Examples
     --------
+
+    **a. Handling various input formats**
```
Nit: Don't need to add the `a.`, `b.` `c.` at the beginning. IMO the bolding makes each section distinct

smarie replied:
On I'll do this

mroeschke commented on pandas/core/tools/datetimes.py:
```diff
@@ -854,16 +914,93 @@ def to_datetime(
+
+    **c. Timezones and time offsets**
+
+    The default behaviour (``utc=False``) might be confusing concerning timezones:
```
Don't think this sentence adds much. I think the bullets above each example are sufficient

smarie replied:
This is to oppose the first group of bullets (utc=False) with the end of the subsection (utc=True).
Would replacing this with 

    `The default behaviour (``utc=False``) is as follows:`

be better ?

mroeschke replied:
Yes that would be better thanks

mroeschke commented on pandas/core/tools/datetimes.py:
```diff
@@ -854,16 +914,93 @@ def to_datetime(
+
+    - ...and not if the timezone-naive elements are strings
+
+    >>> pd.to_datetime(["2020-01-01 01:00 -01:00", "2020-01-01 03:00"])
```
I would say this is a bug actually. This should match the above example. Could you open and issue and remove this example for now?

smarie replied:
I created #44949

mroeschke reviewed (changes requested):
A lot better, thank you. Thanks for your persistence with this PR!

jreback reviewed (changes requested):
wow, looking good will review after @mroeschke comments

Commit:
Changed as per code review

smarie commented:
thanks for taking the time to review ! I took into account all of your comments @mroeschke , let me know if the result still needs modification.

Commit:
Changed as per code review

Commit:
what's new attempt

smarie commented:
I also realize that I did not write an entry in what's new. I see in 1.1.0 that "clarified documentation" can be inserted in the bugfix section (see https://pandas.pydata.org/docs/whatsnew/v1.1.0.html?highlight=clarified#missing), so added something similar, let me know.

mroeschke commented on doc/source/whatsnew/v1.4.0.rst:
```diff
@@ -637,8 +637,8 @@ Timedelta
 Timezones
 ^^^^^^^^^
 - Bug in :func:`to_datetime` with ``infer_datetime_format=True`` failing to parse zero UTC offset (``Z``) correctly (:issue:`41047`)
+- Clarified :func:`to_datetime` documentation concerning parameter ``utc`` and the impact of its default value (``False``) on parsing datetimes from a timezone with varying time offsets (daylight savings) (:issue:`42229`).
```
We don't add whatsnew entries for doc changes anymore.

smarie replied:
ok

mroeschke reviewed (commented):
Looks pretty good! Off to @jreback

jreback reviewed (changes requested):
@smarie looks ok to me, if you can remove the whatsnew new and ping on green

jreback commented:
@smarie if you can update and merge master

Commit:
Revert "what's new attempt"

This reverts commit 70a7c8f0e41395b9c0813b892249bd50c1e607f7.

Commit:
Merge branch 'master' of https://github.com/pandas-dev/pandas into patch-1

smarie commented:
Sorry for the delay @jreback @mroeschke . I updated the branch and removed the whats new entry.
Build seems ok except for a few that seem unrelated to this PR

jreback commented:
thanks @smarie if you can rebase one more time pls.

also if you can go over the text again and be consistent about using backticks vs class refs (e.g. i think we prefer to use ```:class:`Timestamp` ``` over ``` ``Timestamp`` ``` (and same for `datetime.datetime`); you'l likely have to render this locally to make sure things come out ok. (this is also ok as a followup, but since need rebase anyhow....)

jreback commented:
cc @mroeschke if comments

mroeschke reviewed (approved):
LGTM after @jreback's comments about the rendering consistency. CI failure will probably be unrelated.

Commit:
Merge branch 'master' of https://github.com/pandas-dev/pandas into patch-1

Commit:
Changed as per code review: added sphinx directives wherever possible, instead of the double backticks.

Commit:
Changed as per code review: added const role

Commit:
sphinx role

Commit:
Changed as per code review: sphinx roles

Commit:
minor change again

jreback commented:
from the doc build: 

> /home/runner/work/pandas/pandas/pandas/core/tools/datetimes.py:docstring of pandas.core.tools.datetimes.to_datetime:72: WARNING: Unexpected indentation.

Commit:
Last polishing round: sphinx roles and a few fixes

smarie commented on pandas/core/tools/datetimes.py:
```diff
@@ -697,46 +697,47 @@ def to_datetime(
-        Specify a date parse order if `arg` is str or its list-likes.
-        If True, parses dates with the day first, eg 10/11/12 is parsed as
-        2012-11-10.
+        Specify a date parse order if `arg` is str or its list-like.
```
```suggestion
        Specify a date parse order if `arg` is str or is list-like.
```

Commit:
Fixed typo

smarie commented:
Thanks, It seems that this is now fixed @jreback . I fixed as much as possible the various sphinx roles everywhere, using explicit `:const:` whenever it was relevant.

smarie commented:
The remaining error is "/home/runner/work/pandas/pandas/pandas/core/ops/array_ops.py
  /home/runner/work/pandas/pandas/pandas/core/ops/array_ops.py:98:17 - error: Unnecessary "cast" call; type is already "dtype[Unknown]" (reportUnnecessaryCast)"

so it seems unrelated

jreback commented:
thanks @smarie very nice!

smarie commented:
Thanks very much @jreback and @mroeschke for your patience and careful review.
